apm-sniffer/optional-plugins/customize-enhance-plugin/src/main/java/org/apache/skywalking/apm/plugin/customize/interceptor/BaseInterceptorMethods.java

```diff
@@ -54,33 +54,44 @@ void beforeMethod(Method method, Object[] allArguments) {
                     operationNameSuffix.append(CustomizeExpression.parseExpression(expression, context));
                 }
             }
-            if (tags != null && !tags.isEmpty()) {
-                for (Map.Entry<String, String> expression : tags.entrySet()) {
-                    spanTags.put(expression.getKey(), CustomizeExpression.parseExpression(expression.getValue(), context));
-                }
-            }
-            if (logs != null && !logs.isEmpty()) {
-                for (Map.Entry<String, String> entries : logs.entrySet()) {
-                    String expression = logs.get(entries.getKey());
-                    spanLogs.put(entries.getKey(), CustomizeExpression.parseExpression(expression, context));
-                }
-            }
+            evalAndPopulate(context, tags, spanTags);
+            evalAndPopulate(context, logs, spanLogs);
+
             operationName = operationNameSuffix.insert(0, operationName).toString();
 
             AbstractSpan span = ContextManager.createLocalSpan(operationName);
-            if (!spanTags.isEmpty()) {
-                for (Map.Entry<String, String> tag : spanTags.entrySet()) {
-                    span.tag(Tags.ofKey(tag.getKey()), tag.getValue());
-                }
-            }
-            if (!spanLogs.isEmpty()) {
-                span.log(System.currentTimeMillis(), spanLogs);
-            }
+            tagSpanTags(span, spanTags);
+            tagSpanLogs(span, spanLogs);
         }
     }
 
-    void afterMethod(Method method) {
-        ContextManager.stopSpan();
+    void afterMethod(Method method, Object ret) {
+        if (!ContextManager.isActive()) {
+            return;
+        }
+
+        if (ret == null) {
+            ContextManager.stopSpan();
+            return;
+        }
+
+        Map<String, Object> configuration = CustomizeConfiguration.INSTANCE.getConfiguration(method);
+        Map<String, Object> context = CustomizeExpression.evaluationReturnContext(ret);
+
+        Map<String, String> tags = MethodConfiguration.getTags(configuration);
+        Map<String, String> spanTags = new HashMap<String, String>();
+        Map<String, String> logs = MethodConfiguration.getLogs(configuration);
+        Map<String, String> spanLogs = new HashMap<String, String>();
+
+        evalReturnAndPopulate(context, tags, spanTags);
+        evalReturnAndPopulate(context, logs, spanLogs);
+
+        final AbstractSpan localSpan = ContextManager.activeSpan();
```

Comment:
All variables here(configuration, context, tags, spanTags, localSpan) could share with before method by using plugin V2 APIs(https://skywalking.apache.org/docs/skywalking-java/v8.8.0/en/setup/service-agent/java-agent/java-plugin-development-guide/#v2-apis-1)

**MethodInvocationContext context** is the key point to this. You could save a lot resources and GC cost by sharing this.

Reply:
Updated